src/app/study-filters/study-filters.component.ts

```diff
@@ -0,0 +1,101 @@
+})
+export class StudyFiltersComponent extends StatefulComponent implements OnInit, OnDestroy {
+  @Input() public dataset: Dataset;
+
+  @Validate(SetNotEmpty, {message: 'Select at least one.'})
+  public selectedStudies: Set<string> = new Set<string>(['']); // Used by the child component(Treelist checkbox)
+  public allStudyNodes: DatasetNode[] = null;
+  public datasetSubscription: Subscription;
+
+  @ViewChild('nav') public ngbNav: NgbNav;
+
+  public constructor(protected store: Store, public datasets: DatasetsService) {
+    super(store, StudyFiltersState, 'studyFilters');
+  }
+
+  public showStudyFilters = false;
+
+  public ngOnInit(): void {
+    super.ngOnInit();
+    const datasets$ = this.datasets.getDatasetsObservable();
+    const datasetTrees = new Array<DatasetNode>();
+    this.datasetSubscription = datasets$.subscribe(datasets => {
+      datasets.filter(d => !d.parents.length).map(d => {
+        datasetTrees.push(new DatasetNode(d, datasets));
+      });
+      this.allStudyNodes = this.getAllStudyNodesById(datasetTrees, this.dataset.id);
+    });
+    this.store.selectOnce(state => state.studyFiltersState as StudyFiltersModel).subscribe(state => {
+      // restore state
+      if ((state?.studyFilters?.length ?? 0) !== 0) {
+        const timer = setTimeout(() => {
+          if (this.showStudyFilters === false) {
+            this.showStudyFilters = true;
+            this.ngbNav.select('studies');
+            this.updateSelectedStudies(new Set(state.studyFilters));
+            clearInterval(timer);
+          }
+        }, 200);
+      }
+    });
+  }
+
+  public ngOnDestroy(): void {
+    this.datasetSubscription.unsubscribe();
+  }
+
+  public onNavChange(): void {
+    if (this.ngbNav.activeId === 'studies') {
+      this.selectedStudies.add(''); // Otherwise the selected studies are going to be empty, preventing navigation
+    } else {
+      this.selectedStudies.clear();
+    }
+    this.store.dispatch(new SetStudyFilters([]));
+    this.store.dispatch(new StateReset(StudyFiltersState));
+  }
+
+  private getAllStudyNodesById(datasetTrees: DatasetNode[], datasetId: string): DatasetNode[] {
+    const result: DatasetNode[] = [];
+    for (const dataset of datasetTrees) {
+      if (dataset.dataset.id === datasetId) {
+        result.push(dataset);
+      } else if (dataset.children && dataset.children.length > 0) {
+        result.push(...this.getAllStudyNodesById(dataset.children, datasetId));
+      }
+    }
+    return result;
+  }
+
+  public updateSelectedStudies(newValues: Set<string>): void {
+    const stateValues = new Set<string>();
+    this.selectedStudies = newValues;
+    newValues.forEach(value => {
+      if (findNodeById(this.allStudyNodes[0], value).children.length === 0) {
+        stateValues.add(value);
+      }
+    });
+    this.store.dispatch(new SetStudyFilters(Array.from(stateValues)));
+  }
+
+  public updateFilters(): void {
+    this.updateSelectedStudies(this.selectedStudies);
+  }
```

Comment:
There is more potential for simplifying the code here. The important call here is the one that dispatches a new state to the store - we can easily get rid of the rest of the code. Here's how this can be done:

1. Delete both of these methods and create a new one called `updateState` which simply dispatches a new state to the store
2. Remove line `53`, manually set `this.selectedStudies` to whatever was being passed to the function on that line and add a call to our new `updateState` method
3. Modify `TreelistCheckboxComponent` so that you **do not** add/delete parent datasets to `this.selectedStudies` when checking/unchecking their checkboxes. Always add only the leaf nodes (i.e. the studies, not the datasets) to `this.selectedStudies`.